Marlin/src/module/temperature.cpp

```diff
@@ -981,9 +981,9 @@ void Temperature::min_temp_error(const heater_id_t heater_id) {
         static float temp_iState[HOTENDS] = { 0 },
                      temp_dState[HOTENDS] = { 0 };
         static bool pid_reset[HOTENDS] = { false };
-        const float hotend_ee = temp_hotend[ee].celsius,
-                    hotend_target = temp_hotend[ee].target,
-                    pid_error = hotend_target - hotend_ee;
+        const celsius_t hotend_ee = temp_hotend[ee].celsius,
```

Comment:
Those are here to compute the fixed point to floating point only once. If you change them to celsius_t and use cast below, it'll cost the conversion multiple time.

Reply:
I'm surprised the compiler in `-O2` would do something different … unless that is one of the useful side-effects of wrapping the temperature value in a struct/class.

I am still not fond of over-complicating `celsius_t`. The convenience of the nice simple type we currently have might just outweigh the added inconveniences of an over-designed complex type.

All of Marlin's internal temperature work is done with celsius units, and there is no confusion about that. Any conversion from some other unit is done immediately on input, and afterward will only be converted to another unit for display. So there is no need to create a series of temperature types. So single simple type, perhaps able to convert to and from other units as part of the type definition, is all that is actually needed. Making the type overly-strict is not a Requirement.

As a coder who works with these types every day I would much prefer that type conversions were done as transparently as they were before, and that we did not require any extra gesticulations to get a celsius temperature in the form of a primitive type. We simply don't need all these extras attached to the type.

One of the elements of Marlin's "coding style" has been to keep it informal and non-pedantic. Keeping the `celsius_t` a simple alias for `int16_t` provides maximum convenience, and makes it much easier to manage and use the value. That factor alone outweighs formalism — or at least incomplete formalism.

When I introduced the `types.h` file with all those lengthy position types, it demonstrated just how much work and detail is required to make custom types behave in "the way that you would want and expect" at any given use. If you're expecting the fixed-to-float conversions here to be un-optimizable by the compiler then maybe the type needs more work. It really should behave transparently and not introduce more inconveniences.

Reply:
There's no difference between `celsius_t` being a `int16_t` or a `fixp`. If you convert it to float in the code at multiple places, the compiler might do it once (and store the value in a register) or do it multiple time if it does not find a free register or estimate it's better this way. When I do it myself, I'm preventing the compiler to decide on its own. 
It's probable -O2 or -Os will give the same code but it can change any time depending on the code content, I haven't checked.

Converting from int to float on a CPU without a FPU is costly (it requires a recursive division). Doing it with the fixed point type requires a multiplication and a shift (better but still expensive).  

The only difference with a POD and a struct is that the conversion is implicit in the former case and I've made it explicit in the latter case. I can make it implicit, but I think it's better if the developer choose to convert so he's aware of it.

For now, I'll continue with explicit until it's compiling fine on all platform (so I can fix the code with clear intentions) and then switch back to implicit conversion for convenience.

Reply:
Costly only weigh against size when there is a bottleneck. We don't do enough temperature conversion between types to consider it a meaningful bottleneck, so it is okay if it is a little inefficient.

Reply:
There is of course the potential to convert all floating-point in Marlin into fixed-point, knowing the predictable ranges of certain parameters. That would be the most thoroughgoing approach, and to eschew `float` altogether when working with temperatures.

The earliest version of Prusa Firmware converted the planner to use fixed-point because it is known to have a certain limited range of bed size and micro-steps. No one has yet made that conversion in Marlin because it's a bunch of meticulous and specialized work, and it would need to be able to pick the most appropriate type for machines that have larger coordinate ranges or higher numbers of micro-steps.

Reply:
I've chosen 10.6 because it allowed +511 -512 range (and 0.15°C of resolution). We could add an offset here if we decided to increase the resolution (like a FP9.7 + 192°C, to get: +448/-64°C and 0.007° of resolution). If we need more, we can choose a 32 bits type here, with: 12.20 here (it's a #ifdef away to change), so we'd get +4095/-4096 and 1µ°C of resolution, but the cost will be higher for multiplications.